From 79c2351c07edbf3a30547b3f1f3d0f4ebaa50d13 Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Wed, 2 Oct 2024 12:19:44 +0200
Subject: [PATCH 1270/1695] core: pta: test bisect implementation

Add a bisect test sequence to OP-TEE selftest support in the test PTA.

The test checks bisect operation for present and absent cells, on
arrays from 0 to 65 cells.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 53b9fe05ee082b7efb7727d586aabb7369f21fe3 core: pta: test bisect implementation]
Change-Id: I281da151eb9f6d5191e9f325e26a84f7e0f9593f
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/408383
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/pta/tests/misc.c | 88 ++++++++++++++++++++++++++++++++++++++++++-
 1 file changed, 87 insertions(+), 1 deletion(-)

diff --git a/core/pta/tests/misc.c b/core/pta/tests/misc.c
index 63dbea48b..6b04f92c9 100644
--- a/core/pta/tests/misc.c
+++ b/core/pta/tests/misc.c
@@ -3,6 +3,7 @@
  * Copyright (c) 2014, STMicroelectronics International N.V.
  */
 #include <assert.h>
+#include <bisect.h>
 #include <config.h>
 #include <kernel/dt_driver.h>
 #include <malloc.h>
@@ -551,6 +552,91 @@ static int self_test_nex_malloc(void)
 }
 #endif
 
+static int cmp_int(const void *a, const void *b)
+{
+	const int *ia = a;
+	const int *ib = b;
+
+	return CMP_TRILEAN(*ia, *ib);
+}
+
+static int self_test_bisect(void)
+{
+	int array[65] = { 0 };
+	size_t array_sz = 0;
+	int *result = NULL;
+	size_t hide = 0;
+	size_t cnt = 0;
+	int target = 0;
+
+	/*
+	 * Create an array of integers from 0 to n and use bisect to
+	 * find each cell.
+	 * Test out of bound value by bisecting for values -1 and n+1.
+	 * Test not-found cases by hiding a value. It is replaced by
+	 * its neighbor value or by n+2 if there is no neighbor.
+	 */
+	LOG("bisect tests on a %zu element array", sizeof(array));
+
+	for (array_sz = 0; array_sz < ARRAY_SIZE(array); array_sz++) {
+		for (hide = 0; hide < array_sz + 1; hide++) {
+			/* Fill array with incremental values */
+			for (cnt = 0; cnt < array_sz; cnt++)
+				array[cnt] = cnt;
+
+			/* Possibly hide 1 element so that it can't be found */
+			if (array_sz && hide < array_sz) {
+				if (!hide) {
+					if (array_sz == 1)
+						array[hide] = array_sz + 2;
+					else
+						array[hide] = array[hide + 1];
+				} else if (hide == array_sz - 1) {
+					if (array_sz == 1)
+						array[hide] = array_sz + 2;
+					else
+						array[hide] = array[hide - 1];
+				} else {
+					array[hide] = array[hide - 1];
+				}
+			}
+
+			/* Try to find present/out-of-bound/absent cells */
+			for (target = -1; target < (int)array_sz + 1;
+			     target++) {
+				result = bisect_equal(array, array_sz,
+						      sizeof(int), &target,
+						      cmp_int);
+
+				if (target == -1 ||
+				    target == (int)array_sz ||
+				    target == (int)hide) {
+					if (result) {
+						LOG("Found unexpected target %d (size %zu, hide %zu)",
+						     target, array_sz, hide);
+						return -1;
+					}
+				} else {
+					assert(result && *result == target);
+					if (!result) {
+						LOG("Failed to find target %d (size %zu, hide %zu)",
+						     target, array_sz, hide);
+						return -1;
+					}
+					if (*result != target) {
+						LOG("Wrongly found %d for target %d (size %zu, hide %zu)",
+						     *result, target, array_sz,
+						     hide);
+						return -1;
+					}
+				}
+			}
+		}
+	}
+
+	return 0;
+}
+
 /* exported entry points for some basic test */
 TEE_Result core_self_tests(uint32_t nParamTypes __unused,
 		TEE_Param pParams[TEE_NUM_PARAMS] __unused)
@@ -558,7 +644,7 @@ TEE_Result core_self_tests(uint32_t nParamTypes __unused,
 	if (self_test_mul_signed_overflow() || self_test_add_overflow() ||
 	    self_test_sub_overflow() || self_test_mul_unsigned_overflow() ||
 	    self_test_division() || self_test_malloc() ||
-	    self_test_nex_malloc()) {
+	    self_test_nex_malloc() || self_test_bisect()) {
 		EMSG("some self_test_xxx failed! you should enable local LOG");
 		return TEE_ERROR_GENERIC;
 	}
-- 
2.47.2

